### PR TITLE
[2.x] Dataset scopes

### PR DESCRIPTION
**Pull request by fabio-ivona**
This PR will implement **scoped datasets**

in v1 datasets can be defined in 3 ways:

1. inside a test file 
2. inside .php files in a `Datasets` directory placed it the test suite root
3. inside a `Datasets.php` file placed in the test suite root

with this PR datasets can be scoped with the following rules:

1. _a dataset defined inside a test file_ (using the `dataset()` function) **is scoped to the file itself** and will not be accessible from other test files 
2. _a dataset defined inside a test file_ **will have precedence** with datasets with the same name defined outside of the test file
3. _a dataset defined in a Datasets.php file_ will be scoped to all test defined in the same directory and/or child directories
4. _a dataset defined in a .php file in a Datasets directory_ will be scoped to all tests defined in directory where the Datasets directory belongs and all tests in any child directory
5. two datasets with the same name can't share the same scope

given these rules, each test file will see the datasets in its scope, with the nearest taking precedence to the farthest

**Overlapping datasets**

with this directory tree:

```
/tests
   |--Datasets
   |      |--Numbers.php (with a 'numbers' dataset inside)   [1]
   |
   |--Datasets.php (with a 'numbers' dataset inside)         [2]
   |--Pest.php

```
  
-  [1] and [2] will throw an exception as the two `numbers` datasets will have the same scope (/tests)

**Scope precedence**

```
/tests
   |--Datasets
   |      |--Numbers.php (with a 'numbers' dataset inside)   [1]
   |
   |--Feature
   |      |--Controllers
   |      |      |--ApiControllers
   |      |      |     |--PostApiController  (will use `numbers` dataset [2])
   |      |      |     
   |      |      |--Datasets.php  (with a 'numbers' dataset inside) [2]
   |      |      |
   |      |      |--PostController.php (will use `numbers` dataset [2])
   |      |
   |      |--Models
   |      |      |--Datasets
   |      |      |      |--Numbers.php  with a 'numbers' dataset inside) [3]
   |      |      |
   |      |      |--PostController.php (will use `numbers` dataset [3])
   |      |
   |      |--Actions
   |             |--SavePostTest.php   (will use `numbers` dataset [1])
   |--Pest.php

```

**TODO**

- [x] It should be avoided to define tests inside Datasets files, I have to find a way to prevent that (maybe disabling `test` and `it` functions during the `BootFiles` step

### CODE REVIEW

File: src/Functions.php [hunk 1]
```diff
@@ -60,7 +61,17 @@ function beforeEach(Closure $closure = null): BeforeEachCall
      */
     function dataset(string $name, Closure|iterable $dataset): void
     {
-        DatasetsRepository::set($name, $dataset);
+        $file = Backtrace::datasetsFile();
+        $filename = Str::afterLast($file, DIRECTORY_SEPARATOR);
+        $scope = Str::beforeLast($file, DIRECTORY_SEPARATOR);
+
+        if (Str::afterLast($scope, DIRECTORY_SEPARATOR) === 'Datasets') {
```
Comment (lukeraymonddowning) [inline src/Functions.php]: We do these same `Str` checks in multiple files. I wonder if we should introduce a `Support\Dataset` class?

File: src/Repositories/DatasetsRepository.php [hunk 2]
```diff
@@ -67,9 +69,10 @@ public static function has(string $filename, string $description): bool
      */
     public static function get(string $filename, string $description)
     {
+        // dump("requesting file: " . $filename);
```
Comment (lukeraymonddowning) [inline src/Repositories/DatasetsRepository.php]: Left this in by mistake?

File: src/Repositories/DatasetsRepository.php [hunk 3]
```diff
@@ -132,23 +135,53 @@ public static function resolve(string $description, array $dataset): array|null
         return $namedData;
     }
 
+    /**
+     * @return Closure|iterable<int|string, mixed>
+     */
+    private static function getScopedDataset(string $name, string $currentTestFile)
+    {
+        $matchingDatasets = array_filter(self::$datasets, function (string $key) use ($name, $currentTestFile) {
+            [$datasetScope, $datasetName] = explode('>>>', $key);
```
Comment (lukeraymonddowning) [inline src/Repositories/DatasetsRepository.php]: I don't really like that we carry this '>>>' string around everywhere. Should we extract it to a constant?

File: src/Repositories/DatasetsRepository.php [hunk 4]
```diff
@@ -132,23 +135,53 @@ public static function resolve(string $description, array $dataset): array|null
...
+            }
+
+            if (! str_starts_with($currentTestFile, $datasetScope)) {
+                return false;
+            }
+
+            return true;
+        }, ARRAY_FILTER_USE_KEY);
+
+        $closestScopeDatasetKey = array_reduce(array_keys($matchingDatasets), function ($keyA, $keyB) {
```
Comment (lukeraymonddowning) [inline src/Repositories/DatasetsRepository.php]:
```suggestion
        $closestScopeDatasetKey = array_reduce(
            array_keys($matchingDatasets), 
            fn ($keyA, $keyB) => $keyA !== null && strlen($keyA) > strlen($keyB) ? $keyA : $keyB
        );
```

File: src/Exceptions/DatasetAlreadyExist.php [hunk 5]
```diff
@@ -17,8 +17,8 @@ final class DatasetAlreadyExist extends InvalidArgumentException implements Exce
     /**
      * Creates a new Exception instance.
      */
-    public function __construct(string $name)
+    public function __construct(string $name, string $scope)
```
Comment (lukeraymonddowning) [inline src/Exceptions/DatasetAlreadyExist.php:20]: @nunomaduro shall we use the v2 opportunity to rename this to `DatasetAlreadyExists`? Grammatically correct :-)

File: src/Repositories/DatasetsRepository.php [hunk 6]
```diff
@@ -132,23 +135,53 @@ public static function resolve(string $description, array $dataset): array|null
...
         foreach ($datasets as $index => $data) {
             $processedDataset = [];
 
             if (is_string($data)) {
-                if (! array_key_exists($data, self::$datasets)) {
-                    throw new DatasetDoesNotExist($data);
-                }
-
-                $datasets[$index] = self::$datasets[$data];
+                $datasets[$index] = self::getScopedDataset($data, $currentTestFile);
```
Comment (lukeraymonddowning) [inline src/Repositories/DatasetsRepository.php:151]: Small thing, but can the `getScopedDataset` method go below this method?

**Review by lukeraymonddowning — CHANGES_REQUESTED** [review]
Looking pretty sweet so far!

File: src/Functions.php (hunk 1 above)
Comment (fabio-ivona) [reply src/Functions.php]: @lukeraymonddowning I refactored this a bit, extracted to a support class and changed Str:: support functions with dirname and basename, so there is less code to maintain

you can find the changes here: 

https://github.com/pestphp/pest/pull/584/commits/80854d5f8700a958be28433cd60c2de74404c4a3

File: src/Repositories/DatasetsRepository.php (hunk 2 above)
Comment (fabio-ivona) [reply src/Repositories/DatasetsRepository.php]: lol, thanks :rofl:

File: src/Repositories/DatasetsRepository.php (hunk 3 above)
Comment (fabio-ivona) [reply src/Repositories/DatasetsRepository.php]: done

File: src/Repositories/DatasetsRepository.php (hunk 6 above)
Comment (fabio-ivona) [reply src/Repositories/DatasetsRepository.php:151]: done

File: src/Exceptions/DatasetAlreadyExist.php (hunk 5 above)
Comment (fabio-ivona) [reply src/Exceptions/DatasetAlreadyExist.php:20]: It's an internal class, so don't think there are any issues in doing it

File: src/Functions.php (hunk 1 above)
Comment (lukeraymonddowning) [reply src/Functions.php]: So much nicer ❤️

**Review by nunomaduro — APPROVED** [review]
Good job!

**Pull request merged**